task_manager/tests/test_user.py

```diff
@@ -0,0 +1,238 @@
+    assert ServiceUser.objects.count() == 1
+    assert response.status_code == 302
+
+
+def test_edit_user_get_user_logged_in(
+    client: Client, service_user: ServiceUser
+) -> None:
+    # given
+    user_id = service_user.id
+    route = f"/users/{user_id}/edit/"
+
+    # when
+    client.force_login(service_user)
+    response = client.get(route)
+    csrf_token = response.context.get("csrf_token")
+    form = ServiceUserForm(instance=service_user)
+    context = {"form": form, "user_id": user_id, "csrf_token": csrf_token}
+
+    # then
+    expected_html = TemplateResponse(
+        response.wsgi_request, "update_user.html", context=context
+    ).render()
+    assert expected_html.content == response.content
+    assert "Изменение пользователя" in response.content.decode()
+    assert response.status_code == 200
+
+
+def test_edit_user_get_user_not_logged_in(
+    client: Client, service_user: ServiceUser
+) -> None:
+    # given
+    user_id = service_user.id
+    route = f"/users/{user_id}/edit/"
+
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 302
+
+
+def test_edit_user_get_requestuser_is_not_objectuser(
+    client: Client, service_user: ServiceUser
+) -> None:
+    # given
+    client.force_login(service_user)
+    object_user_id = ServiceUser.objects.create_user("username", "", "password").id
+    route = f"/users/{object_user_id}/edit/"
+
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 302
+
+
+def test_edit_user_objectuser_not_exists(
+    client: Client, service_user: ServiceUser
+) -> None:
+    # given
+    client.force_login(service_user)
+    object_user_id = 2
+    route = f"/users/{object_user_id}/edit/"
+
+    # when
+    response = client.get(route)
+
+    # then
+    assert response.status_code == 404
+
+
+@freeze_time("15.07.2024 17:00")
+def test_edit_user_post(client: Client, service_user: ServiceUser) -> None:
+    # given
+    client.force_login(service_user)
+    user_id = service_user.id
+    route = f"/users/{user_id}/edit/"
+    form_data = {
+        "username": "username",
+        "password1": "password",
+        "password2": "password",
+    }
+
+    # when
+    response = client.post(route, data=form_data)
+
+    # then
+    assert ServiceUser.objects.get(id=1).username == "username"
+    assert response.status_code == 302
+
+
+@freeze_time("15.07.2024 17:00")
+def test_edit_user_post_form_not_valid(
+    client: Client, service_user: ServiceUser
+) -> None:
+    # given
+    client.force_login(service_user)
+    user_id = service_user.id
+    user_username = service_user.username
+    route = f"/users/{user_id}/edit/"
+    form_data = {
+        "username": "",
+        "password1": "",
+        "password2": "",
+    }
+
+    # when
+    response = client.post(route, data=form_data)
+
+    # then
+    assert ServiceUser.objects.get(id=1).username == user_username
+    assert response.status_code == 200
```

Comment:
why does it return `200_OK` when validation has failed?

Reply:
cause the same form is refreshed/renewed if form data is not valid